jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathMappings.java

```diff
@@ -231,24 +295,48 @@ public boolean put(PathSpec pathSpec, E resource)
         switch (pathSpec.getGroup())
         {
             case EXACT:
-                String exact = pathSpec.getPrefix();
-                while (exact != null && !_exactMap.put(exact, entry))
+                if (pathSpec instanceof ServletPathSpec)
+                {
+                    String exact = pathSpec.getDeclaration();
+                    while (exact != null && !_exactMap.put(exact, entry))
+                    {
+                        _exactMap = new ArrayTernaryTrie<>((ArrayTernaryTrie<MappedResource<E>>)_exactMap, 1.5);
+                    }
+                }
+                else
                 {
-                    _exactMap = new ArrayTernaryTrie<>((ArrayTernaryTrie<MappedResource<E>>)_exactMap, 1.5);
+                    // This is not a Servlet mapping, turn off optimization on Exact
+                    _optimizedExact = false;
                 }
                 break;
             case PREFIX_GLOB:
-                String prefix = pathSpec.getPrefix();
-                while (prefix != null && !_prefixMap.put(prefix, entry))
+                if (pathSpec instanceof ServletPathSpec)
+                {
+                    String prefix = pathSpec.getPrefix();
+                    while (prefix != null && !_prefixMap.put(prefix, entry))
+                    {
+                        _prefixMap = new ArrayTernaryTrie<>((ArrayTernaryTrie<MappedResource<E>>)_prefixMap, 1.5);
+                    }
+                }
+                else
                 {
-                    _prefixMap = new ArrayTernaryTrie<>((ArrayTernaryTrie<MappedResource<E>>)_prefixMap, 1.5);
+                    // This is not a Servlet mapping, turn off optimization on Prefix
+                    _optimizedPrefix = false;
```

Comment:
Again, we should be able to extract the prefix from any PREFIX_GLOB and put that in the Trie.
So again, for now add a TODO to optimize later.

Reply:
It's not quite as easy as you think.
The comments added show what needs to be resolved.
TODO added as well.